poetry/puzzle/provider.py

```diff
@@ -478,12 +478,11 @@ def complete_package(
             if self._env and not dep.marker.validate(self._env.marker_env):
                 continue
 
-            if (
-                dep.is_optional()
-                and dep.name not in optional_dependencies
-                and not package.is_root()
-            ):
-                continue
+            if not package.is_root():
+                if (
+                    dep.is_optional() and dep.name not in optional_dependencies
+                ) or not set(dep.in_extras).issubset(package.dependency.extras):
```

Comment:
Hi, thanks for working on the issue I reported! I don't know much about the inner workings of `poetry`, but i was curious about the fix and had a look. If I'm interpreting this correctly, you now also skip a dependency if the set of extras in which it's declared isn't a subset of the extras flagged for installation.

If my interpretation is correct, I believe this should rather be 
```suggestion
                ) or not set(dep.in_extras).intersection(package.dependency.extras):
```
Case in point: package `C` declared in both extras `one` and `two` of package `B`, but only `B[one]` declared as a dependency for `A`. This should cause `C` to be selected.

If I completely misunderstood, feel free to dismiss my comment.

Reply:
@michamos I do not fully follow the scenario, just to make sure; does this case cover it?

https://github.com/python-poetry/poetry/pull/3046/commits/205a1796243f3e00d0b1ff042a93b91d87231b12

Reply:
I don't think it does. I'm not sure how to express it in the tests directly, so I'll write it down using `pyproject.toml` syntax. What I mean is the following for `B`:
```toml
[tool.poetry]
name = "B"

[tool.poetry.dependencies]
C = { version = "1.0", optional = true }

[tool.poetry.extras]
one = ["C"]
two = ["C"]
```
then have the `pyproject.toml` for `A` contain
```toml
B = { version = "1.0", extras = ["one"] }
```

I believe your code won't select C as a dependency as `bool({"one", "two"}.issubset({"one"})) is False`.

Reply:
So, this case? https://github.com/python-poetry/poetry/pull/3046/commits/50dc9a09da7d7ac75223d29602cb8c1af458b2e0#diff-4e7cde4641210ec491c396893ff273d0R509